### PR TITLE
add remarks in TransactionAttributeUsage

### PR DESCRIPTION
**Pull request by bettybao1209**
close https://github.com/neo-project/neo/issues/1493

### CODE REVIEW

File: src/neo/Network/P2P/Payloads/TransactionAttribute.cs [hunk 1]
```diff
@@ -17,7 +17,7 @@ void ISerializable.Deserialize(BinaryReader reader)
             Usage = (TransactionAttributeUsage)reader.ReadByte();
             if (!Enum.IsDefined(typeof(TransactionAttributeUsage), Usage))
                 throw new FormatException();
-            Data = reader.ReadVarBytes(252);
+            Data = reader.ReadVarBytes(ushort.MaxValue);
```
Comment (shargon) [inline src/neo/Network/P2P/Payloads/TransactionAttribute.cs]: I think that it's very big, any use case for this size?

```suggestion
            Data = reader.ReadVarBytes(byte.MaxValue);
```

**Comment by erikzhang** [comment]
The URL can be removed?

**Comment by Tommo-L** [comment]
>The URL can be removed?

Oracle will not use `TransactionAttributeUsage.URL` ?

**Comment by erikzhang** [comment]
I think no.

File: src/neo/Network/P2P/Payloads/TransactionAttribute.cs (hunk 1 above)
Comment (bettybao1209) [reply src/neo/Network/P2P/Payloads/TransactionAttribute.cs]: I think 255 bytes may not be enough for remarks, also to keep consistent with the neo2.

**Comment by shargon** [comment]
Oracles will have his own attribute, I think that we can remove it.
Remember that currently there are no syscall for read these attributes @bettybao1209

File: src/neo/Network/P2P/Payloads/TransactionAttribute.cs (hunk 1 above)
Comment (shargon) [reply src/neo/Network/P2P/Payloads/TransactionAttribute.cs]: 64kb for one attribute seems very huge

**Comment by ProDog** [comment]
I have tested it passed.

image
image

image

File: src/neo/Network/P2P/Payloads/Transaction.cs [hunk 2]
```diff
@@ -311,7 +311,8 @@ public StackItem ToStackItem(ReferenceCounter referenceCounter)
                 SystemFee,
                 NetworkFee,
                 ValidUntilBlock,
-                Script,
+                new Array(Attributes.Select(p=> (StackItem)p.ToArray())),
```
Comment (erikzhang) [inline src/neo/Network/P2P/Payloads/Transaction.cs]: I don't think the attributes should be available in smart contracts.

File: src/neo/Network/P2P/Payloads/TransactionAttributeUsage.cs [hunk 3]
```diff
@@ -2,6 +2,23 @@ namespace Neo.Network.P2P.Payloads
...
+        Remark6 = 0xf6,
+        Remark7 = 0xf7,
+        Remark8 = 0xf8,
+        Remark9 = 0xf9,
+        Remark10 = 0xfa,
+        Remark11 = 0xfb,
+        Remark12 = 0xfc,
+        Remark13 = 0xfd,
+        Remark14 = 0xfe,
+        Remark15 = 0xff
```
Comment (erikzhang) [inline src/neo/Network/P2P/Payloads/TransactionAttributeUsage.cs]: Do we really need so many remarks? Maybe one is enough.

File: src/neo/Network/P2P/Payloads/Transaction.cs (hunk 2 above)
Comment (bettybao1209) [reply src/neo/Network/P2P/Payloads/Transaction.cs]: On some level maybe it's reasonable to provide attributes to sc developers for custom use. In some scenario, project parties can use different usage types to distinguish the tx type.

Comment (erikzhang) [reply src/neo/Network/P2P/Payloads/Transaction.cs]: But them are serialized. You can't deserialize them in a contract. So it's useless.

**Comment by erikzhang** [comment]
Maybe we should remove all the attributes. I think remarks can be put in the script.

**Comment by shargon** [comment]
For oracles, it won't be needed the attributes if we use Version.

**Comment by vncoelho** [comment]
I agree with @erikzhang, that would make TX's simpler and the procedure more standard.

**Comment by Tommo-L** [comment]
> Maybe we should remove all the attributes. I think remarks can be put in the script.

Please leave some creative space for users, keep at least one attribute for user-define. For example, users  want to send a little paper on blockchain, we can just set the attribute enough, which don't need to write the contact.

**Comment by shargon** [comment]
> Please leave some creative space for users, keep at least one attribute for user-define. For example, users want to send a little information(like love letter, flag, leave message, etc) on blockchain, we can just set the attribute enough, which don't need to write the contact script.

But what it's the difference with the script?

**Comment by vncoelho** [comment]
For me the script is more clear, @Tommo-L. Let's see what @igormcoelho thinks.

**Comment by igormcoelho** [comment]
I agree with @Tommo-L , in leaving a creative space, that's how many people use blockchain technology for. Regarding doing that on Attribute or Script, maybe it's even better on Script @Tommo-L... one can launch a Notification, which is better registered than an attribute. What do you think?
Or should we have some special  Interop just for random Record (System.Record "my message").

**Comment by erikzhang** [comment]
`Runtime.Log`

**Comment by igormcoelho** [comment]
You're right @erikzhang we can use log... but just to make it more official, is `Log` intended for contract debug logs, or to log nice information? I mean, if Neo has a specific command to Log **nice** stuff, people can use it to validate weddings, certificates, and all nice stuff that grow around blockchain ecosystem. By having a specific **thing** for that, blockchain explorers can treat this information with bigger care... as far as I know, `Log` is just ignored by explorers nowadays (just left for contract devs to analyse themselves).

**Comment by Tommo-L** [comment]
I think it's a big difference between script and attribute. For attribute, you can get the information directly. But for script, you must execute the script and extract log indirectly.

Here is an example, if user want to transfer NEO to another people, he/she can add the comment information. If we add the information in attribute, wallets can display the information directly. But if we use script, wallets show the program **log** for customers, will it be a little weird? Who knows how many logs, and which log is the comments.

**Comment by igormcoelho** [comment]
@Tommo-L nice comment.. I was trying to think on something that could be statically checked, without actual script execution... for example:
PUSH "my nice message"
Runtime.Record

When wallets parse (not execute) this script (on Entry), they are already able to detect the message. However, they would inform this even if it's inside an if:
```
// now in programmble format (random language)
if(xxxx)
    Runtime.Record("mynicemessage")
```
This would be displayed even if execution flow doesn't enter the if.

**Comment by Tommo-L** [comment]
> When wallets parse (not execute) this script (on Entry)

It may add more works on wallets, which may need a compiler to parse the script or vm to execute.

**Comment by erikzhang** [comment]
The wallet should execute the script, or it must connect to a server that executes the script.

**Comment by Tommo-L** [comment]
> or it must connect to a server that executes the script.

I know. It'll be much easier and clearer for wallet to display the comment information for customer by transaction attribute.

**Comment by shargon** [comment]
In fact `Attributes` could be more powerful in the future, and if you don't want to use it, you only waste 1 byte (for the length).

**Comment by Tommo-L** [comment]
It's controversial and may have some bad effects. We will close the PR.

**Pull request closed without merging**